tao_compiler/mlir/disc/transforms/disc_dynamic_offloading.cc

```diff
@@ -0,0 +1,433 @@
+  SmallVector<Value, 4> dynShape;
+  for (size_t i = 0; i < type.getRank(); i++) {
+    if (type.getShape()[i] == ShapedType::kDynamic) {
+      dynShape.push_back(b.create<memref::DimOp>(loc, refBuffer, i));
+    }
+  }
+  auto allocOp = b.create<memref::AllocOp>(loc, type, dynShape);
+  StringRef attrName = SymbolicDimOp::getSymbolicDimAttrName();
+  if (refBuffer.getDefiningOp()->hasAttr(attrName)) {
+    allocOp.getOperation()->setAttr(
+        attrName, refBuffer.getDefiningOp()->getAttr(attrName));
+  }
+  return allocOp.getResult();
+}
+
+// InsertRematBlock create a remat block for the living buffer:
+// reload and offload blocks are always pair in graph:
+//
+// if remat_cond:
+//   return offload(buffer)
+// else:
+//   return dummy_buffer
+// ......
+// if remat_cond:
+//   return reload(buffer)
+// else:
+//   return buffer
+void DiscOffloadingPass::InsertRematBlock(mlir::OpBuilder& b,
+                                          LivingBuffer& livingBuffer,
+                                          Value rematCond,
+                                          std::vector<OpWithPosition>& ops,
+                                          int64_t minSymValue) {
+  // TODO(yancey): we need a custom operator to handle the dummy buffer to avoid
+  // call alloc or dealloc operator
+  auto buffer = livingBuffer.buffer;
+  auto startOp = livingBuffer.start;
+  auto endOp = livingBuffer.end;
+  StringRef attrName = SymbolicDimOp::getSymbolicDimAttrName();
+  auto deviceMemrefType = buffer.getType().cast<MemRefType>();
+  auto hostMemrefType = MemRefType::get(
+      deviceMemrefType.getShape(), deviceMemrefType.getElementType(), {}, 0);
+  // insert offload block
+  auto offloadIfOp =
+      b.create<scf::IfOp>(startOp->getLoc(),
+                          /*resultTypes*/ hostMemrefType, rematCond,
+                          /*hasElseRegion*/ true);
+  offloadIfOp.getOperation()->setAttr(
+      attrName, buffer.getDefiningOp()->getAttr(attrName));
+  offloadIfOp.getOperation()->setAttr(kRematBlockTypeAttr,
+                                      b.getStringAttr("offload"));
+  offloadIfOp.getOperation()->setAttr(kRematMinSymDim,
+                                      b.getI64IntegerAttr(minSymValue));
+  offloadIfOp.getThenRegion().front().clear();
+  b.setInsertionPointToEnd(&offloadIfOp.getThenRegion().front());
+  auto hostBuffer = cloneBuffer(b, endOp->getLoc(), buffer);
+  hostBuffer.setType(MemRefType::get(deviceMemrefType.getShape(),
+                                     deviceMemrefType.getElementType(), {}, 0));
+  b.create<lmhlo_disc::D2HOp>(endOp->getLoc(), buffer, hostBuffer);
+  b.create<memref::DeallocOp>(endOp->getLoc(), buffer);
+  b.create<scf::YieldOp>(endOp->getLoc(), hostBuffer);
+
+  offloadIfOp.getElseRegion().front().clear();
+  b.setInsertionPointToStart(&offloadIfOp.getElseRegion().front());
+  auto dummyHostBuffer = cloneBuffer(b, endOp->getLoc(), buffer);
+  dummyHostBuffer.getDefiningOp()->setAttr(kRematBufferAttr,
```

Comment:
will this dummy host buffer cause some unexpected behavior .... if it's used somewhere